src/libsyntax/parse/parser.rs

```diff
@@ -245,6 +245,23 @@ enum PrevTokenKind {
     Other,
 }
 
+// Simple circular buffer used for keeping few next tokens.
+#[derive(Default)]
+struct LookaheadBuffer {
+    buffer: [TokenAndSpan; LOOKAHEAD_BUFFER_CAPACITY],
+    start: usize,
+    end: usize,
+}
+
+const LOOKAHEAD_BUFFER_CAPACITY: usize = 8;
```

Comment:
`x % LOOKAHEAD_BUFFER_CAPACITY` is faster.
Not much difference either way, though.